### PR TITLE
FEATURE: provide node-type and fusion prototype declarations as symbols

### PR DESCRIPTION
**Pull request by PRGfx**
Node-types and fusion prototype declarations can be reached through the "go to symbol" dialog.

(Prototypes sadly can't be found when using their namespace. I maybe have something wrong when getting the name from the `FusionPrototypeSignature`)

### CODE REVIEW

File: src/main/java/de/vette/idea/neos/search/NodeTypeContributor.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    private static class NodeTypeNavigationItem extends de.vette.idea.neos.search.NavigationItem {
+
+        /**
+         * Creates a new display item.
+         *
+         * @param psiElement The PsiElement to navigate to.
+         * @param text       Text to show for this element.
+         */
+        public NodeTypeNavigationItem(@NotNull NavigatablePsiElement psiElement, @NotNull String text) {
+            super(psiElement, text, ICON_NEOS);
```
Comment (cvette) [inline src/main/java/de/vette/idea/neos/search/NodeTypeContributor.java:62]: you can use `NeosIcons.NODE_TYPE`  here

File: src/main/java/de/vette/idea/neos/search/NodeTypeContributor.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        @Override
+        public ItemPresentation getPresentation() {
+            return new de.vette.idea.neos.search.NavigationItem.FlowSourceItemPresentation(
```
Comment (cvette) [inline src/main/java/de/vette/idea/neos/search/NodeTypeContributor.java:67]: No need for a fully-qualified name here

**Review by cvette — CHANGES_REQUESTED** [review]
Awesome! Didn't even think of this :)

**Comment by PRGfx** [comment]
Another thing I'm really not happy with is how the path is split into package name and the "relevant" part within each package.
I think it would be extremely handy to have some utility function like
* `@Nullable String NeosUtil::getPackageName(VirtualFile)`
* `String NeosUtil::getPathInPackage(VirtualFile)` (defaults to path in project)
* `boolean NeosUtil::isTestFile(VirtualFile)` I would really like if the symbol contributor could highlight entries from test-directories, so I know that this is probably not the declaration I'm looking for (like the [ColoredItemPresentation](https://upsource.jetbrains.com/idea-ce/file/idea-ce-4b94ba01122752d7576eb9d69638b6e89d1671b7/platform/core-api/src/com/intellij/navigation/ColoredItemPresentation.java) I think).

(Possibly with overloads for `PsiFile`s or path strings)
However with all these things I'm not sure how best to tackle them: `ResolveEngine::findPackagesDirectory` for example will only find the main `Packages` directory, but most people will have some `DistributionPackages` directory (or something named differently). For the tests it might work to test for `^[tT]ests?\/` in the package-relative path.

**Comment by PRGfx** [comment]
> > `ResolveEngine::findPackagesDirectory` for example will only find the main `Packages` directory, but most people will have some `DistributionPackages` directory (or something named differently).
> 
> When using the local repository pattern, wouldn't you still work within the Packages directory and exclude DistributionPackages? At least that's how we do it.

I had some awkward behavior with git while working through the symlinked directory, so I always set it up the other way around :see_no_evil: 
I didn't notice any trouble with this plugin so far in that regard.

> But I agree, there is some room for improvement. One way to do it would be to make it configurable via the settings. Using "Packages" as default, for convenience.

Maybe evaluating the repository definitions of type path in the root composer.json might be feasible.

> > For the tests it might work to test for ^[tT]ests?/ in the package-relative path.
> 
> That might be be a use case for scopes: https://www.jetbrains.com/help/idea/configuring-scopes-and-file-colors.html#creating-a-new-custom-scope

Great point, haven't utilized that so far! I would have expected that a plugin would be able to provide scope configurations like it can provide internal live templates, but I haven't found anything yet – anyway that's a different topic then.


In my eyes that leaves this PR with two things:
* I now changed how prototype symbols are found (3cdc39e). That improves searching for them, as the namespace can be included. However it seems a bit clunky in the `FusionPrototypeDeclarationContributor::processElementsWithName`, but I'm not sure how to properly access the index.
* I'm not completely happy with the ItemPresentation locationString. It currently tries to see what's the package and then returns something like `[Package.Name] Components/Something.fusion` or `[Package.Name] Development/NodeTypes.Something.yaml`. I like having the package name quite prominently so I can see whether something is an override and the path should be the important part in most cases, however it falls apart for i.e. test fixtures outside of `Resources/Private/Fusion`..
  (apparently I haven't even found how to correctly get the project-relative path of a VirtualFile ^^)

**Comment by cvette** [comment]
> Great point, haven't utilized that so far! I would have expected that a plugin would be able to provide scope configurations like it can provide internal live templates, but I haven't found anything yet – anyway that's a different topic then.

I did some research and it turns out there is an extension point that let's you implement a custom filter for test files: https://upsource.jetbrains.com/idea-ce/file/idea-ce-b8f393732fae6beb1e4f70ec4d605f2838253529/platform/projectModel-api/src/com/intellij/openapi/roots/TestSourcesFilter.java

**Comment by PRGfx** [comment]
Unfortunately I let this get a bit stale..
* Regarding the test sources highlighting, I think it could be a nice addition, but obviously another feature.
* Regarding the topic of displaying the symbol location and finding/guessing the associated package, I lean towards some project-utilities to do that (maybe involve looking for composer files of type ` neos-*` and [extract the package name](https://github.com/neos/flow-development-collection/blob/c62b8f3e29387e8f588dc3e1f62fa1b60c8b8ece/Neos.Flow/Classes/Package/PackageManager.php#L878)) outside the scope of this PR.
  While digging a bit I found the `SymbolPresentationUtil.getFilePathPresentation(file)` which I would use rather than the mess I currently had. As the fallback path anyways. I currently had this construct in there where I would tell the `NavigationItem` some path part at which it would then split the file path and retrieve the path "relevant" for fusion and node-type configs respectively.
  As I said earlier it's quite fragile as well..
  @cvette do you have thoughts on that before I throw out my current implementation and just return the file-path for now?
* lastly the "requested change" – I "resolved" the discussion bits as GitHub labelled them "outdated", but that probably was not the right move :see_no_evil:

**Comment by cvette** [comment]
> While digging a bit I found the SymbolPresentationUtil.getFilePathPresentation(file) which I would use rather than the mess I currently had. As the fallback path anyways. I currently had this construct in there where I would tell the NavigationItem some path part at which it would then split the file path and retrieve the path "relevant" for fusion and node-type configs respectively.
As I said earlier it's quite fragile as well..
@cvette do you have thoughts on that before I throw out my current implementation and just return the file-path for now?

I would keep it simple for now. As you mentioned, packages should better be handled by a service with proper caching, but that is for another time.

Can you please clean up the commits a bit when you"re done? Then, we can get this merged :)

**Pull request merged**